rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+    },
+    {
+      "type": "enum MyError",
+      "concreteTypeId": "44781f4b1eb667f225275b0a1c877dd4b9a8ab01f3cd01f8ed84f95c6cd2f363",
+      "metadataTypeId": 1
+    },
+    {
+      "type": "u8",
+      "concreteTypeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+    }
+  ],
+  "metadataTypes": [
+    {
+      "type": "(_, _)",
+      "metadataTypeId": 0,
+      "components": [
+        {
+          "name": "__tuple_element",
+          "typeId": 2
+        },
+        {
+          "name": "__tuple_element",
+          "typeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+        }
+      ]
+    },
+    {
+      "type": "enum MyError",
+      "metadataTypeId": 1,
+      "components": [
+        {
+          "name": "A",
+          "errorMessage": "error A",
+          "typeId": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d"
+        },
+        {
+          "name": "B",
+          "errorMessage": "error B",
+          "typeId": 0
+        }
+      ]
+    },
+    {
+      "type": "u64",
+      "metadataTypeId": 2
+    }
+  ],
+  "functions": [
+    {
+      "inputs": [
+        {
+          "name": "a",
+          "concreteTypeId": "c89951a24c6ca28c13fd1cfdc646b2b656d69e61a92b91023be7eb58eb914b6b"
+        }
+      ],
+      "name": "main",
+      "output": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d",
+      "attributes": null
+    }
+  ],
+  "loggedTypes": [
+    {
+      "logId": "4933727799282657266",
+      "concreteTypeId": "44781f4b1eb667f225275b0a1c877dd4b9a8ab01f3cd01f8ed84f95c6cd2f363"
+    }
+  ],
+  "messagesTypes": [],
+  "configurables": [],
+  "errorCodes": {
+    "18446744069414584320": {
+        "pos": {
+          "file": "main.sw",
+          "line": 13,
+          "col": 10
+        },
+        "logId": null,
+        "msg": "a str panic"
+    },
+    "18446744069414584321": {
+        "pos": {
+          "file": "main.sw",
+          "line": 14,
+          "col": 10
+        },
+        "logId": null,
+        "msg": "another str panic"
+    },
+    "18446744069414584322": {
+        "pos": {
+          "file": "main.sw",
+          "line": 15,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    },
+    "18446744069414584323": {
+        "pos": {
+          "file": "main.sw",
+          "line": 16,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    }
+    "18446744069414584324": {
+        "pos": {
+          "file": "main.sw",
+          "line": 17,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    },
+    "18446744069414584325": {
+        "pos": {
+          "file": "main.sw",
+          "line": 18,
+          "col": 10
+        },
+        "logId": "4933727799282657266",
+        "msg": null,
+    }
+  }
```

Comment:
These four all have the same `logId` with `msg: null`, the only difference being the position in the file. The first two are `MyError::A` and the second two are `MyError::B(..)`. I can't see how, from this specification, the SDKs would be able to determine which error variant corresponds to these error codes.

Also, this looks like every instance of `MyError::A` will have a different code, which I'm not sure is desirable. For example in unit tests, it would be nice to know that `MyError::A` will always have the same code, so that we can use `should_panic = <code>` to assert that `MyError::A` was thrown.

On that note, it would be even better if we can make `should_panic` support the literal type like `should_panic = MyError::A`, and `should_panic = MyError::B(10, 10)`, but that might be out of scope here. To support it in the future, though, we'd need to lay the groundwork by having a consistent error code per error variant type within a given sway program.

Reply:
>I can't see how, from this specification, the SDKs would be able to determine which error variant corresponds to these error codes

The SDK doesn't have to figure out which variant from the code because this information is already available in the log: the encoded error contains the enum discriminator.

>every instance of MyError::A will have a different code

This is deliberate. There's no other way to know where the invocation is, and I think this information is important.

The VM can't do this because it does not have access to information about the source code.

Reply:
> it would be even better if we can make `should_panic` support the literal type like `should_panic = MyError::A`

This would be great usability! And as suggested even having not only the variant name, but also concrete parameters. I think this is all possible within the proposed in this RFC. In the sense that all the building blocks and needed information is there.

Since the code will be assigned by the compiler and visible only in the ABI, I would see the `should_panic = <code>` as a fallback, ideally never needed to be used.

Reply:
I'm going to close this since the info is available, but I do think using that feature of naming the errors for tests would be a very useful addition.